homeassistant/components/apprise/notify.py

```diff
@@ -66,6 +70,26 @@ def send_message(self, message="", **kwargs):
         However, if any tags are specified, then they will be applied
         to the notification causing filtering (if set up that way).
         """
+
         targets = kwargs.get(ATTR_TARGET)
         title = kwargs.get(ATTR_TITLE, ATTR_TITLE_DEFAULT)
-        self.apprise.notify(body=message, title=title, tag=targets)
+
+        data = kwargs.get(ATTR_DATA)
+        attachments = data.get(ATTR_ATTACHMENTS) if data else None
+
+        # Default Attachment
+        attach = None
+        if attachments:
+            # Support file attachments (such as security camera images,
+            # documents, icons, etc). The attachments can be a list and/or
+            # a string.  The AppriseAttachment() object accepts both
+            attach = apprise.AppriseAttachment()
+            attach.add(attachments)
+
+            # Iterate over detected attachments to verify that they have been
+            # whitelisted.
+            attach.attachments = [
+                a for a in attach if a and self.hass.config.is_allowed_path(a.path)
```

Comment:
It seems weird that you first call `attach.add`, then filter the list on there. Better to not add any attachment that is never white listed, or even instantiate `AppriseAttachment` object if none are whitelisted.

Reply:
The thing with apprise, is it takes URL's as input for attachments (so that you can pass along options too) such as:
- `file:///tmp/live-camera.jpeg?cache=3`
- `http://home.security.hkvision/path/to/live/image.jpeg`
- `http://dropbox.com/path/to/content?dl=1`

A path (relative or absolute) `/path/to/file` with **AppriseAttachment.add()** actually gets interpreted as `file:///path/to/file` internally.

Apprise doesn't actually attempt to access the attachments until a notification (call to `notify()`) is made, or if you actually attempt to access the attachment itself.

The code i added forces all attachments to be retrieved early (web requests get fetched and stored in whatever the Python `tmpdir` path is set to.  By referencing the `.path`, it is explicitly getting the full path to the file.  The logic works because I'm not sure if your `is_allowed_path()` would catch something like: `file://etc/passwd?cache=40` where as checking the already parsed `attachment.path` would explicitly return `/etc/passwd` (and get caught by your whitelist).

The other reason to check after the `add()` call is because the function itself accepts both a single string, or a list of strings (and does all of the handling/error checking of this up front).

Hopefully what I'm explaining makes sense :slightly_smiling_face: 

Thoughts?

Reply:
It seems extremely dangerous that apprise will literally fetch any url from inside Home Assistant and send it out to an external device!

Reply:
Apprise won't send content anywhere; it will only retrieve. Think of a use case where you need to connect to a remote camera to grab it's latest image (as an example). You can fetch the image and pass it along through when an event is triggered.  Local files are referred to no differently than anything else. It's just the `file://` (and others) prefix that is additionally supported by Apprise. To provide an absolute/relative path check in the format your white-list function requires/expects, the _Apprise Attachment URL_ just needs to be pre-parsed.  Hence  `AppriseAttachment.path` is this very reference.

Reply:
> Apprise won't send content anywhere

It's being attached to the outgoing notification..

Reply:
Sorry, I misinterpreted your reference to being that it downloaded stuff and send it elsewhere (outside of the notification). Yes you're correct.  But how would it be dangerous?

I can tell you that:
- It only retrieves the (external/local) content based on the configuration you set (which is local and maintained by the operator).
- Content is removed from the `{python_temp_path}` after the `AppriseAttachment` goes out of scope (so only the life of the `hass.notify()` call.)
- As mentioned already: local files are just referenced _as is_ (just like the other notification services do).
- Restrictions can be set; so right now it's [set to only allow up to 5MB files](https://github.com/caronc/apprise/blob/68fcfafc863c7175e2c10834d2a380e57805a5e0/apprise/attachment/AttachBase.py#L63).

Thanks to your original observation, all files are passed through your white-list function (i didn't realize that existed until your review).

To me, it opens a fantastic new world of supporting all kinds of external devices that you can grab content from prior to the sending of the notification.  If you had 6 cameras around your house, your attachment could include all 6 images in the attachment.  Depending on which service you're using (Discord, Slack, etc), Apprise will make sure all attachments make it to the destination.

Reply:
Since a hacker can set the target too, it means that if it can call services it can now fetch any local url (even ones that require authentication) and send the data to themselves on any of the supported platforms.

Reply:
But if the system is already compromised, then the hacker already have the access to this information anyway (if it's being hosted as you suggest). How does Apprise make this worse; it's merely offering attachments just like the other notification services. It's no different then what is happening here with [Home Assistant Slack attachments](https://www.home-assistant.io/integrations/slack/) which can be externally referenced sources? Also, Apprise can't fetch local files outside of your _whitelist_ as per latest observation you showed.

Reply:
If the Slack integration also allows fetching any local url from Home Assistant and sent it out, we should remove that too. 

A normal user can not manage the system but can send notifications, and can now expose any internal available url.

Reply:
> If the Slack integration also allows fetching any local url from Home Assistant and sent it out, we should remove that too.

Personally, I think this seems a bit extreme. You're taking away some fantastic functionality that Home Assistant has on the account of some of the bad apples out there who may abuse it.  May I suggest you leave it but instead maybe we could update the white-listing to also do hostnames as well?  Don't allow anything that resolves to 'localhost' or perhaps the same (or a new) whitelist? Perhaps the Apprise and Slack components (_there may be more_) could be updated to additionally check against this new list?

Reply:
@balloob : I was just seeing [the HASS notification services](https://companion.home-assistant.io/en/notifications/attachments) that might be suffering from the same flaw i'm trying to introduce here. I haven't check into the source code yet, but is documented as:

> Configuration
>  -    url (Required): The URL of content to use as the attachment. This URL must be accessible from the Internet, _or the receiving device must be on the same network as the hosted content._

Would it suffice if i blacklisted all attachment URLs by default unless the configuration had an entry that looked like this:

```
# Example configuration.yaml entry using URLs
notify:
  - platform: apprise
    url: YOUR_APPRISE_URLS
    whitelist: [
        "myserver.localhost",
        "jacks.server.localhost:9000"
   ]
```

Attachments that are not of type `file://` (same as `/a/path/to/location/already/whitelist-checked`) would be checked against this list above?
* If no `whitelist` is specified then all URLs would be blocked.
* Possibly support regular expressions?
    * The entry `server*` would become `server.*` behind the scenes.
    * All hostnames would be escaped, so other then `*` and maybe `?` (for single match).

This will allow people like me to just set an entry like:
```
# Example configuration.yaml entry using URLs
notify:
  - platform: apprise
    url: YOUR_APPRISE_URLS
    whitelist: [
        "*",
   ]
```
   
For ease, the white list can be a single string too:
```
# Example configuration.yaml entry using URLs
notify:
  - platform: apprise
    url: YOUR_APPRISE_URLS
    whitelist: "*"
```

Thoughts?

Reply:
It's ok to send any url to the user per notification. It is not ok, under any circumstance ever, to have Home Assistant download the url and send the content as part of the notification.

Reply:
I understand, I love Home Assistant to much to try to change your mind and anger you :slightly_smiling_face: 

I guess it would be best to simply close this pull request and it's [related documentation](https://github.com/home-assistant/home-assistant.io/pull/11461) PR as well.  While I'm disappointed, I truly do respect your opinion.  Going forward I'll just keep the notification aspect of Apprise going as is, and not the attachment support (with respect to the HASS integration of it).